src/setup/testnet/mod.rs

```diff
@@ -0,0 +1,199 @@
+//! Utilities for setting up a testnet consisting of 3 nodes.
+
+use std::{
+    fmt,
+    fmt::Write,
+    fs, io,
+    net::{IpAddr, SocketAddr},
+    path::{Path, PathBuf},
+};
+
+use crate::setup::{
+    config::{CONFIG_FILE, DEFAULT_PORT, ZIGGURAT_CONFIG},
+    node::Node,
+};
+
+/// Testnet's directory for nodes' configs.
+const TESTNET_DIR: &str = "testnet";
+
+/// Validators file name.
+const VALIDATORS_FILE_NAME: &str = "validators.txt";
+
+/// A struct allowing for convenient starting and stopping a small testnet.
+pub struct TestNet {
+    // Setup information for each node. Used for writing configuration.
+    setups: [NodeSetup; 3],
+    // Running nodes. Used to stopping the testnet.
+    running: Vec<Node>,
+    // Sets whether to log the node's output to Ziggurat's output stream.
+    use_stdout: bool,
+}
+
+impl Default for TestNet {
+    fn default() -> Self {
+        Self {
+            setups: [
+                NodeSetup::new(
+                    "127.0.0.1".parse().unwrap(),
+                    "nHUSqn9qjEF7JJkVqvY7BFLMKdqP5KLLEjo5oB4QH43ADDndRawB".into(),
```

Comment:
where do these come from? shouldn't they be some `const`s (e.g. `const validator_keys: [&str; 3] = [...]`) for better visibility?

Reply:
Could be. It's only the question whether we want stuff to be split horizontally or vertically.

Reply:
whatever `cargo fmt` prefers :stuck_out_tongue:

Reply:
Done